src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1762,31 +1913,137 @@ public sealed class One : Tests
                 public override void Filter(
                     DecisionDagBuilder builder,
                     BoundDagTest test,
+                    DagState state,
                     IValueSet? whenTrueValues,
                     IValueSet? whenFalseValues,
                     out Tests whenTrue,
                     out Tests whenFalse,
                     ref bool foundExplicitNullTest)
                 {
+                    if (Test is BoundDagEvaluation)
+                    {
+                        // there cannot be any implication for an evaluation node.
+                        whenTrue = whenFalse = this;
+                        return;
+                    }
+
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        precondition: out Tests precondition,
+                        sideeffect: out Tests sideeffect,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    Debug.Assert(sideeffect is True or One(BoundDagAssignmentEvaluation));
+
+                    // Given:
+                    //
+                    //  - "test" as a test that has already occurred,
+                    //  - "other" as a subsequent test,
+                    //  - S as a possible side-effect (expected to always evaluate to True),
+                    //  - and P as a pre-condition under which we need to evaluate S,
+                    //
+                    // we proceed as follows:
+                    //
+                    //  - If "test" being true proves "other" to be also true, we rewrite "other" as ((P && S) || other),
+                    //    Because we have determined that on this branch, "other" would always succeed if the pre-condition is met.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be reduced to True which means "other" is insignificant.
+                    //
+                    //  - If "test" being true proves "other" to be false, we rewrite "other" as (!(P && S) && other),
```

Comment:
nit: either I was confused earlier, or some re-ordering happened. The comment cases here don't match the order below. In the code, the order is currently "test implies true other", "unrelated", "test implies false other".
Consider renaming `decisionPermitsTrueOther` to `decisionImpliesFalseOther` (and inverting it's value) for readability.

Reply:
To avoid unnecessary noise I can invert the condition with  `!decisionPermitsTrueOther`, does that work for you?

Reply:
Sure